src/ocf_freelist.c

```diff
@@ -0,0 +1,381 @@
+				freelist_size);
+
+		if (!freelist_size) {
+			/* init empty freelist and move to next one */
+			freelist->part[freelist_idx].head =
+					collision_table_entries;
+			freelist->part[freelist_idx].tail =
+					collision_table_entries;
+			continue;
+		}
+
+		/* find first invalid cacheline */
+		phys = next_phys_invalid(cache, phys);
+		ENV_BUG_ON(phys == collision_table_entries);
+		idx = ocf_metadata_map_phy2lg(cache, phys);
+		++phys;
+
+		/* store freelist head */
+		freelist->part[freelist_idx].head = idx;
+
+		/* link freelist elements using partition list */
+		prev = collision_table_entries;
+		while (--freelist_size) {
+			phys = next_phys_invalid(cache, phys);
+			ENV_BUG_ON(phys == collision_table_entries);
+			next = ocf_metadata_map_phy2lg(cache, phys);
+			++phys;
+
+			ocf_metadata_set_partition_info(cache, idx,
+					PARTITION_INVALID, next, prev);
+
+			prev = idx;
+			idx = next;
+
+			OCF_COND_RESCHED_DEFAULT(step);
+		}
+
+		/* terminate partition list */
+		ocf_metadata_set_partition_info(cache, idx, PARTITION_INVALID,
+			collision_table_entries, prev);
+
+		/* store freelist tail */
+		freelist->part[freelist_idx].tail = idx;
+	}
+
+	/* we should have reached the last invalid cache line */
+	phys = next_phys_invalid(cache, phys);
+	ENV_BUG_ON(phys != collision_table_entries);
+
+	env_atomic64_set(&freelist->total_free, num_free_clines);
+}
+
+static void ocf_freelist_add_cache_line(ocf_freelist_t freelist,
+		uint32_t ctx, ocf_cache_line_t line)
+{
+	struct ocf_cache *cache = freelist->cache;
+	struct ocf_part *freelist_part = &freelist->part[ctx];
+	ocf_cache_line_t tail;
+	ocf_cache_line_t line_entries = ocf_metadata_collision_table_entries(
+							freelist->cache);
+	ocf_part_id_t invalid_part_id = PARTITION_INVALID;
+
+	ENV_BUG_ON(line >= line_entries);
+
+	if (env_atomic64_read(&freelist_part->curr_size) == 0) {
+		freelist_part->head = line;
+		freelist_part->tail = line;
+
+		ocf_metadata_set_partition_info(cache, line, invalid_part_id,
+				line_entries, line_entries);
+	} else {
+		tail = freelist_part->tail;
+
+		ENV_BUG_ON(tail >= line_entries);
+
+		ocf_metadata_set_partition_info(cache, line, invalid_part_id,
+				line_entries, tail);
+		ocf_metadata_set_partition_next(cache, tail, line);
+
+		freelist_part->tail = line;
+	}
+
+	env_atomic64_inc(&freelist_part->curr_size);
+	env_atomic64_inc(&freelist->total_free);
+}
+
+static bool ocf_freelist_get_cache_line_ctx(ocf_freelist_t freelist,
+		uint32_t ctx, ocf_cache_line_t *cline)
+{
+	if (env_atomic64_read(&freelist->part[ctx].curr_size) == 0)
+		return false;
+
+	ocf_freelist_lock(freelist, ctx);
+
+	if (env_atomic64_read(&freelist->part[ctx].curr_size) == 0) {
+		ocf_freelist_unlock(freelist, ctx);
+		return false;
+	}
+
+	*cline = freelist->part[ctx].head;
+	_ocf_freelist_remove_cache_line(freelist, ctx, *cline);
+
+	ocf_freelist_unlock(freelist, ctx);
+
+	return true;
+}
+
+static int get_next_freelist(ocf_freelist_t freelist)
```

Comment:
consider having independent victim idx per exec ctx

Reply:
This doesn't seem to improve things much. To make this work we would need to be pinned to current execution context for the duration of slow path. Current implementation only pins execution to exec context for fast path.